File: leetcode/golang/39_combination_sum.go
func combinationSum(candidates []int, target int) [][]int {
    sort.Ints(candidates);
    var res [][]int
    var state []int
    var dfs func(candidates []int, target int, start int) 
    dfs = func(candidates []int, target int, start int) {
        if target == 0 {
            res = append(res, append([]int(nil), state...))
            return
        }
        for i := start; i < len(candidates); i++ {
            if target >= candidates[i] {
                state = append(state, candidates[i])
                dfs(candidates, target - candidates[i], i)
                state = state[:len(state) - 1]
            }
        }
    }
    dfs(candidates, target, 0)
    return res
}

File: leetcode/c++/39_combination_sum.cpp
// https://leetcode.cn/problems/combination-sum/description

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        dfs(candidates, target, 0);
        return res;
    }
private:
    vector<vector<int>> res;
    vector<int> state;
    void dfs(vector<int>& candidates, int target, int start) {
        if (target == 0) {
            res.push_back(state);
            return;
        }
        for (int i = start; i < candidates.size(); i++) {
            if (target >= candidates[i]) {
                state.push_back(candidates[i]);
                dfs(candidates, target - candidates[i], i);
                state.pop_back();
            }
        }
    }
};

// 时间复杂度和空间复杂度挺复杂的，估计也不会问